src/agreementManager/AgreementManager.ts

```diff
@@ -104,37 +112,293 @@ class AgreementManager {
+                filteredMilestoneProofs.push(fullResult.milestoneProofs[i]);
+                filteredMilestoneSnapshots.push(
+                    fullResult.milestoneSnapshots[i]
+                );
+            }
+        }
+
+        return {
+            encodedLatestFinalizedState: fullResult.encodedLatestFinalizedState,
+            encodedLatestCorrectState: fullResult.encodedLatestCorrectState,
+            virtualVotingBlocks: filteredVirtualVotingBlocks,
+            milestoneProofs: filteredMilestoneProofs,
+            milestoneSnapshots: filteredMilestoneSnapshots
+        };
+    }
+
     /**
      * Get the latest finalized state and latest signed state (by signer) with virtual voting blocks
+     * This method accounts for exit points as milestones in the new design where participants can exit within a fork
      */
-    public getFinalizedAndLatestWithVotes(
+    public getFinalizedAndLatestWithMilestones(
         forkId: ForkId,
         signerAddress: Address
     ): {
         encodedLatestFinalizedState: Bytes;
         encodedLatestCorrectState: Bytes;
         virtualVotingBlocks: BlockConfirmationStruct[];
+        milestoneProofs: MilestoneProofStruct[];
+        milestoneSnapshots: StateSnapshot[];
     } {
-        const thresholdAddresses = new Set<Address>(
-            this.storage.stateSnapshots.getGenesisSnapshotDataByForkId(forkId)
-                ?.snapshotData.participants as Address[]
-        );
-
-        if (thresholdAddresses.size === 0) {
+        const genesisSnapshot =
+            this.storage.stateSnapshots.getGenesisSnapshotDataByForkId(forkId);
+        if (!genesisSnapshot) {
             throw new Error("Fork not found");
         }
 
+        // Get all exit points for this fork - these are the milestones
+        const exitPoints = this.storage.exitPoints.getExitPointsInRange(forkId);
+
+        // Get all blocks sorted by height descending
+        const blockEntries = Array.from(
+            this.storage.blocks.getBlocksByForkId(forkId, SortOrder.DESC)
+        );
+
         let encodedLatestFinalizedState: Bytes | undefined;
         let encodedLatestCorrectState: Bytes | undefined;
         let virtualVotingBlocks: BlockConfirmationStruct[] = [];
-        let requiredSignatures = new Set<Address>(thresholdAddresses);
+        let milestoneProofs: MilestoneProofStruct[] = [];
+        let milestoneSnapshots: StateSnapshot[] = [];
+
+        // Start with genesis snapshot for the first milestone
+        let currentSnapshot = genesisSnapshot;
+
+        // Process exit points in chronological order - each exit point is a milestone
+        const sortedExitPoints = [...exitPoints].sort(
+            (a, b) => Number(a) - Number(b)
+        );
+
+        for (const exitPointHeight of sortedExitPoints) {
+            const milestoneData = this.processMilestone(
+                exitPointHeight,
+                blockEntries,
```

Comment:
This shouldn't be a range of blocks, but an iterator starting from the first blocks that should be analyzed